src/ckb_chain/funding/funding_tx.rs

```diff
@@ -0,0 +1,253 @@
+    core::{BlockView, Capacity, TransactionView},
+    packed,
+    prelude::*,
+};
+use molecule::{
+    bytes::{BufMut as _, BytesMut},
+    prelude::*,
+};
+
+/// Funding transaction wrapper.
+///
+/// It includes extra fields to verify the transaction.
+#[derive(Clone, Debug, Default)]
+pub struct FundingTx {
+    tx: Option<TransactionView>,
+}
+
+#[allow(dead_code)]
+#[derive(Clone, Debug, Default)]
+pub struct FundingUdtInfo {
+    /// The UDT type script
+    type_script: packed::Script,
+    /// CKB amount to be provided by the local party.
+    local_ckb_amount: u64,
+    /// CKB amount to be provided by the remote party.
+    remote_ckb_amount: u64,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct FundingRequest {
+    /// UDT channel info
+    pub udt_info: Option<FundingUdtInfo>,
+    /// The funding cell lock script args
+    pub funding_cell_lock_script_args: packed::Bytes,
+    /// Assets amount to be provided by the local party
+    pub local_amount: u64,
+    /// Fee to be provided by the local party
+    pub local_fee_rate: u64,
+    /// Assets amount to be provided by the remote party
+    pub remote_amount: u64,
+}
+
+// TODO: trace locked cells
+#[derive(Clone)]
+pub struct FundingContext {
+    pub secret_key: secp256k1::SecretKey,
+    pub rpc_url: String,
+    pub funding_source_lock_script: packed::Script,
+    pub funding_cell_lock_script: packed::Script,
+}
+
+#[allow(dead_code)]
+struct FundingTxBuilder {
+    funding_tx: FundingTx,
+    request: FundingRequest,
+    context: FundingContext,
+}
+
+impl TxBuilder for FundingTxBuilder {
+    fn build_base(
+        &self,
+        _cell_collector: &mut dyn CellCollector,
+        _cell_dep_resolver: &dyn CellDepResolver,
+        _header_dep_resolver: &dyn HeaderDepResolver,
+        _tx_dep_provider: &dyn TransactionDependencyProvider,
+    ) -> Result<TransactionView, TxBuilderError> {
+        let funding_cell = self
+            .build_funding_cell()
+            .map_err(|err| TxBuilderError::Other(err.into()))?;
+        let mut outputs: Vec<packed::CellOutput> = vec![funding_cell.0];
+        let mut outputs_data: Vec<packed::Bytes> = vec![funding_cell.1];
+
+        if let Some(ref tx) = self.funding_tx.tx {
+            for (i, output) in tx.outputs().into_iter().enumerate().skip(1) {
+                outputs.push(output.clone());
+                outputs_data.push(tx.outputs_data().get(i).unwrap_or_default().clone());
+            }
+        }
+
+        let builder = match self.funding_tx.tx {
+            Some(ref tx) => tx.as_advanced_builder(),
+            None => packed::Transaction::default().as_advanced_builder(),
+        };
+        let tx = builder
+            .set_outputs(outputs)
+            .set_outputs_data(outputs_data)
+            .build();
+
+        Ok(tx)
+    }
+}
+
+impl FundingTxBuilder {
+    fn build_funding_cell(&self) -> Result<(packed::CellOutput, packed::Bytes), FundingError> {
+        let remote_funded = self
```

Comment:
I think at this stage, we can safely assure that both parties have already decided which live cells to use to fund the channel. This is because both parties must have go through a lengthy transaction collaboration process and they have already sent a `tx_complete` message. So it seems to me `FundingTxBuilder` must have all the inputs and outputs information of the transaction. It is only the witness of the other party that is missing and we need to set our own witness.

Reply:
FundingTxBuilder is responsoble to collect live cells during negotiation.